peps/pep-0764.rst

```diff
@@ -0,0 +1,334 @@
+comma-separated list of ``key: value`` pairs within braces constructor
+(``{k: <type>}``) is allowed, and should be specified directly as the type
+argument (i.e. it is not allowed to use a variable which was previously
+assigned a :class:`dict` instance).
+
+Inlined typed dictionaries can be referred to as *anonymous*, meaning they
+don't have a name (see the `runtime behavior <Runtime behavior>`_
+section).
+
+It is possible to define a nested inlined dictionary::
+
+    Movie = TypedDict[{'name': str, 'production': TypedDict[{'location': str}]}]
+
+    # Note that the following is invalid as per the updated `type_expression` grammar:
+    Movie = TypedDict[{'name': str, 'production': {'location': str}}]
+
+Although it is not possible to specify any class arguments such as ``total``,
+any :external+typing:term:`type qualifier` can be used for individual fields::
+
+    Movie = TypedDict[{'name': NotRequired[str], 'year': ReadOnly[int]}]
+
+Inlined typed dictionaries are implicitly *total*, meaning all keys must be
+present. Using the :data:`~typing.Required` type qualifier is thus redundant.
+
+Type variables are allowed in inlined typed dictionaries, provided that they
+are bound to some outer scope::
+
+    class C[T]:
+        inlined_td: TypedDict[{'name': T}]  # OK, `T` is scoped to the class `C`.
+
+    reveal_type(C[int]().inlined_td['name'])  # Revealed type is 'int'
+
+
+    def fn[T](arg: T) -> TypedDict[{'name': T}]: ...  # OK: `T` is scoped to the function `fn`.
+
+    reveal_type(fn('a')['name'])  # Revealed type is 'str'
+
+
+    type InlinedTD[T] = TypedDict[{'name': T}]  # OK, `T` is scoped to the type alias.
+
+
+    T = TypeVar('T')
+
+    InlinedTD = TypedDict[{'name': T}]  # Not OK, `T` refers to a type variable that is not bound to any scope.
+
+Typing specification changes
+----------------------------
+
+The inlined typed dictionary adds a new kind of
+:external+typing:term:`type expression`. As such, the
+:external+typing:token:`~expression-grammar:type_expression` production will
+be updated to include the inlined syntax:
+
+.. productionlist:: inlined-typed-dictionaries-grammar
+    new-type_expression: `~expression-grammar:type_expression`
+                       : | <TypedDict> '[' '{' (string: ':' `~expression-grammar:annotation_expression` ',')* '}' ']'
+                       :       (where string is any string literal)
+
+Runtime behavior
+----------------
+
+Although :class:`~typing.TypedDict` is commonly referred as a class, it is
+implemented as a function at runtime. To be made subscriptable, it will be
+changed to be a class.
+
+Creating an inlined typed dictionary results in a new class, so ``T1`` and
+``T2`` are of the same type::
+
+    from typing import TypedDict
+
+    T1 = TypedDict('T1', {'a': int})
+    T2 = TypedDict[{'a': int}]
+
+As inlined typed dictionaries are are meant to be *anonymous*, their
+:attr:`~type.__name__` attribute will be set to an empty string.
```

Comment:
That is going to be confusing, you'll get behavior like this:

```
>>> t = type("", (), {})
>>> t
<class '__main__.'>
>>> t()
<__main__. object at 0x104af1df0>
```

I'd set it to something like "<anonymous TypedDict>".